### ISSUE TITLE
Bootstrapping fails repeatedly

### ISSUE THREAD

**Issue by FlorianHockmann** [scylladb/scylla#8793]
This is Scylla's bug tracker, to be used for reporting bugs only.
If you have a question about Scylla, and not a bug, please ask it in
our mailing-list at scylladb-dev@googlegroups.com or in our slack channel.

- [x] I have read the disclaimer above, and I am reporting a suspected malfunction in Scylla.

*Installation details*
Scylla version (or git commit hash): 4.4.2-0.20210520.93457807b
Cluster size: 3 (should be, but only 1 is `UP` and 1 is bootstrapping, 1 hasn't joined yet again)
OS (RHEL/CentOS/Ubuntu/AWS AMI): Ubuntu 20.04.2 LTS, Scylla running in Docker

*Hardware details (for performance issues)*          Delete if unneeded
Platform (physical/VM/cloud instance type/docker): Docker

I want to give a bit of context first to explain the situation we're now in as it might be relevant for this issue:

We had a cluster of three nodes that was still on Scylla 2.1 and we wanted to follow the recommended upgrade path to get it on version 4.4. Unfortunately we ran into a couple of bugs during the upgrade that were already fixed in newer Scylla versions. When we upgraded to 3.1, the data got apparently even corrupted. We're using JanusGraph on top of Scylla and that suddenly didn't even have our graph schema intact any more, so our Gremlin queries started failing because of some missing property keys.
That is why we decided to start with a completely new cluster and import a snapshot from one node that was created on version 3.0 and therefore still had our data without this problem. We wanted to do it like this:

1. Create a Scylla cluster with just one node on version 3.0 (same version that created the snapshot).
2. Load our CQL schema and set the replication factor to 1.
3. Load the data from the snapshot with `sstableloader`.
4. Update the node to version 4.4.2. (We skipped all versions in between now due to the problem we ran into with version 3.1.)
5. Run `nodetool compact` to update the SSTables to the newest version. (We created a snapshot after this step that we could use if we need to follow a different approach.)
6. Add the other two nodes (also installed with 4.4.2) one after another.
7. Increase the replication factor to 3.
8. Run `nodetool repair` and `nodetool clean` to ensure that all nodes have exactly the data they need.

We're now at step 6 where we try to bootstrap a second node, but it seems like the bootstrapping will not succeed. During the bootstrapping we see these two logs messages regularly:

on the bootstrapping node:
```
Writing a partition with too many rows [[JANUSGRAPH_KEYSPACE]/edgestore:00000008208d9080] (121005 rows)
```

and on the node that already has the data:
```
(rate limiting dropped 255 similar messages) oversized request (length 131072) submitted. dazed and confuzed, trimming its weight from 55296 down to 52947
```

and a few times the bootstrapping seems to fail which apparently causes the bootstrapping Scylla instance to crash and then start with the bootstrapping again afterwards.

The relevant log messages on the boostrapping node seem to be:

```
{"log":"WARN  2021-06-02 22:53:05,534 [shard 6] large_data - Writing a partition with too many rows [janusgraph_grid/edgestore:a000000a86b7c980] (159868 rows)\n","stream":"stderr","time":"2021-06-02T22:53:05.535044454Z"}
{"log":"INFO  2021-06-02 23:19:57,610 [shard 0] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] Received failed complete message, peer=[UpNodeIP]\n","stream":"stderr","time":"2021-06-02T23:19:57.611744266Z"}
{"log":"WARN  2021-06-02 23:19:57,611 [shard 0] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] Streaming plan for Bootstrap-janusgraph_grid-index-2 failed, peers={[UpNodeIP]}, tx=0 KiB, 0.00 KiB/s, rx=76029007 KiB, 34190.61 KiB/s\n","stream":"stderr","time":"2021-06-02T23:19:57.612438446Z"}
{"log":"WARN  2021-06-02 23:19:57,613 [shard 0] range_streamer - Bootstrap with [UpNodeIP] for keyspace=janusgraph_grid failed, took 12672.512 seconds: streaming::stream_exception (Stream failed)\n","stream":"stderr","time":"2021-06-02T23:19:57.61352037Z"}
{"log":"WARN  2021-06-02 23:19:57,613 [shard 0] range_streamer - Bootstrap failed, took 12672 seconds, nr_ranges_remaining=1450\n","stream":"stderr","time":"2021-06-02T23:19:57.614060009Z"}
{"log":"WARN  2021-06-02 23:19:57,613 [shard 0] range_streamer - Bootstrap failed to stream. Will retry in 303 seconds ...\n","stream":"stderr","time":"2021-06-02T23:19:57.614087083Z"}
{"log":"INFO  2021-06-02 23:19:57,666 [shard 29] rpc - client [UpNodeIP]:49693: unexpected eof on a stream while reading data: expected 166 got 97\n","stream":"stderr","time":"2021-06-02T23:19:57.666846331Z"}
{"log":"INFO  2021-06-02 23:19:57,667 [shard 15] rpc - client [UpNodeIP]:59375: unexpected eof on a stream while reading data: expected 162 got 146\n","stream":"stderr","time":"2021-06-02T23:19:57.667936302Z"}
{"log":"ERROR 2021-06-02 23:19:57,669 [shard 29] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] Failed to handle STREAM_MUTATION_FRAGMENTS (receive and distribute phase) for ks=janusgraph_grid, cf=edgestore, peer=[UpNodeIP]: seastar::nested_exception: seastar::rpc::stream_closed (rpc stream was closed by peer) (while cleaning up after seastar::rpc::stream_closed (rpc stream was closed by peer))\n","stream":"stderr","time":"2021-06-02T23:19:57.66979076Z"}
```

and the node that is `UP` only reports at that time that the other node is now `DOWN`:

```
"log":"INFO  2021-06-02 22:42:54,017 [shard 21] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] Start sending ks=janusgraph_grid, cf=edgestore, estimated_partitions=3811537, with new rpc streaming\n","stream":"stderr","time":"2021-06-02T22:42:54.018317187Z"}
{"log":"INFO  2021-06-02 23:19:57,606 [shard 0] gossip - InetAddress [BoostrappingNodeIP] is now DOWN, status = BOOT\n","stream":"stderr","time":"2021-06-02T23:19:57.607125795Z"}
{"log":"INFO  2021-06-02 23:19:57,606 [shard 0] stream_session - stream_manager: Close all stream_session with peer = [BoostrappingNodeIP] in on_dead\n","stream":"stderr","time":"2021-06-02T23:19:57.607202007Z"}
{"log":"WARN  2021-06-02 23:19:57,607 [shard 0] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] Streaming plan for Bootstrap-janusgraph_grid-index-2 failed, peers={[BoostrappingNodeIP]}, tx=76040325 KiB, 34195.76 KiB/s, rx=0 KiB, 0.00 KiB/s\n","stream":"stderr","time":"2021-06-02T23:19:57.6080128Z"}
{"log":"WARN  2021-06-02 23:19:57,610 [shard 0] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] stream_transfer_task: Fail to send to [BoostrappingNodeIP]:0: seastar::rpc::stream_closed (rpc stream was closed by peer)\n","stream":"stderr","time":"2021-06-02T23:19:57.611268989Z"}
{"log":"WARN  2021-06-02 23:19:57,610 [shard 0] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] Failed to send: seastar::rpc::stream_closed (rpc stream was closed by peer)\n","stream":"stderr","time":"2021-06-02T23:19:57.611297406Z"}
{"log":"WARN  2021-06-02 23:19:57,610 [shard 0] stream_session - [Stream #de5258e0-c3f3-11eb-95eb-000000000000] Streaming error occurred, peer=[BoostrappingNodeIP]\n","stream":"stderr","time":"2021-06-02T23:19:57.611462849Z"}
{"log":"INFO  2021-06-02 23:19:57,687 [shard 0] gossip - InetAddress [BoostrappingNodeIP] is now UP, status = BOOT\n","stream":"stderr","time":"2021-06-02T23:19:57.688368766Z"}
```

I attach the log files of both nodes for the relevant time where the error occurred and the bootstrapping started again. If you need the full logs, then I can of course also provide them.

Logs of the bootstrapping node:

[boostrapping-node-excerpt.log](https://github.com/scylladb/scylla/files/6590445/boostrapping-node-excerpt.log)

Logs of the node that streams the data:

[up-node-excerpt.log](https://github.com/scylladb/scylla/files/6590447/up-node-excerpt.log)

We started the bootstrapping already on May 1 and the disk utilization (~1TB) is already higher on the bootstrapping node than on the node that is up, although we expected just around half the utilization, given that we have a replication factor of 1 right now and both nodes should now only be responsible for half the data the up node was responsible for before the bootstrapping.
So it doesn't look like the bootstrapping will succeed at all (or maybe not even terminate until the disk is full?).

I already found #4319 which sounds quite similar but the error messages are different, so I think that this is not a duplicate.

But it was mentioned in #4319 that large partitions could cause such problems and the log message `Writing a partition with too many rows` and `oversized request` seem to hint at this. So I checked it on the node that is up, but it doesn't report any big partitions:

```
cqlsh> SELECT * FROM system.large_partitions;

 keyspace_name | table_name | sstable_name | partition_size | partition_key | compaction_time
---------------+------------+--------------+----------------+---------------+-----------------

(0 rows)
```

and also no large rows or cells.

It would be good to find some solution that allows us to get our cluster in a working state again as it is now already basically down for over a week due to the problems I mentioned. So, if you could suggest a different approach to initialize the cluster that could avoid this problem, then that would be much appreciated. Would it for example make sense to start the cluster fresh again with all three nodes on 4.4.2, a replication factor of 3, and then load a snapshot with `sstableloader` again? Or would that probably result in a similar problem later if we try to perform any streaming operations, like a repair?

**Comment by avikivity** [comment]
I think we want to find the root cause of the problems first, they look very serious. Otherwise it may fail later with even worse consequences.

If we're lucky, these are networking problems caused by docker, but we have to prove that first.

**Comment by FlorianHockmann** [comment]
Thanks for your quick response, @avikivity!

> I think we want to find the root cause of the problems first

OK, understood.

The bootstrapping seems to have finally succeeded after about 55h. There were some more occurrences of the errors I mentioned above but in the end we got a "Bootstrap completed!" and the node is in state `UP` now. Our Gremlin test queries via JanusGraph also still succeed so the data seems to be in an overall valid state.
So we'll now add the 3rd node to the cluster and see whether it will have the same problems or not. I'll of course report back here to give an update on that.
Or do you suspect that we still have bigger problems and shouldn't continue at all with setting up the cluster?

**Comment by asias** [comment]
@FlorianHockmann With 4.5, there is an easier way to load the sstables from snapshot into a new cluster directly with the load and stream feature. You can take a look here:  https://github.com/scylladb/scylla/pull/7846

- Create a cluster with 3 nodes
- Create the same schema for the table
- Put the sstables from the snapshot into the upload dir for the table in question. You can either put all those sstables into a single node or split them and put them into 3 nodes evenly
- Run  curl -X POST "http://{ip}:10000/storage_service/sstables/{keyspace}?cf={table}&load_and_stream=true

**Comment by FlorianHockmann** [comment]
Thanks a lot, @asias! Your last comment is definitely a big relief! So we now know that we don't have bigger problems with our cluster and we should probably only investigate the network problems.

#7846 sounds like a great feature that would have definitely been good to have for us as it would have been much simpler and probably a lot faster to initialize the cluster again. But I think that we will first try the planned approach and add the third node as we now already have a cluster of two nodes where we just need to add one more node and then increase the replication factor to 3. That way we don't have to install a release candidate on our production cluster and start the whole process again of loading a snapshot.

**Comment by asias** [comment]
> Thanks a lot, @asias! Your last comment is definitely a big relief! So we now know that we don't have bigger problems with our cluster and we should probably only investigate the network problems.
> 
> #7846 sounds like a great feature that would have definitely been good to have for us as it would have been much simpler and probably a lot faster to initialize the cluster again. But I think that we will first try the planned approach and add the third node as we now already have a cluster of two nodes where we just need to add one more node and then increase the replication factor to 3. That way we don't have to install a release candidate on our production cluster and start the whole process again of loading a snapshot.

Sure, stick to the current approach is the best choice at the moment. To speed up the bootstrap, you could try adding --compaction-static-shares 10 to the bootstrapping node to reduce compaction to speed up the bootstrap process.

**Comment by FlorianHockmann** [comment]
The bootstrapping unfortunately had the same problem while bootstrapping the third node and this time the bootstrapping didn't recover automatically. We had to stop Scylla manually and start it again to get the bootstrapping to continue which was especially unfortunate in this case as it stopped on a Friday evening so we only restarted it this morning.

For some reason the bootstrapping node didn't seem to complete its shutdown successfully. Last time, the `compaction_manager` logged for each shard that it's being stopped  and afterwards we saw the log message `Shutting down database was successful`, followed by a few more shutdown log messages after which Scylla was restarted automatically (by Docker Compose in our case probably). But this time, there were no log messages after the `compaction_manager` was stopped. The log message `Shutting down database was successful` and all following log messages about the shutdown were missing.
Scylla was also still up until we stopped it manually, e.g., `nodetool status` still responded.


Apart from that, we also saw a few error / warn messages that were not present last time:

On the bootstrapping node:
```
Jun 4, 2021 @ 18:47:21.139 ERROR 2021-06-04 16:47:21,138 [shard 22] rpc - client [UpNodeIP2]:56662: server stream connection dropped: Unknown parent connection 1 on shard 22
```

and on one of the nodes that are already `UP`:

```
Jun 4, 2021 @ 18:50:11.815 [Stream #61771151-c554-11eb-bd86-000000000007] Streaming error occurred, peer=[BoostrappingNodeIP] WARN
Jun 4, 2021 @ 18:50:11.815 [Stream #61771151-c554-11eb-bd86-000000000007] Failed to send: std::runtime_error ([Stream #61771151-c554-11eb-bd86-000000000007] GOT STREAM_MUTATION_DONE from [UpNodeIP1]: Can not find stream_manager) WARN
Jun 4, 2021 @ 18:50:11.815 [Stream #61771151-c554-11eb-bd86-000000000007] stream_transfer_task: Fail to send to [BoostrappingNodeIP]:0: std::runtime_error ([Stream #61771151-c554-11eb-bd86-000000000007] GOT STREAM_MUTATION_DONE from [UpNodeIP1]: Can not find stream_manager) WARN
Jun 4, 2021 @ 18:50:11.815 [Stream #61771151-c554-11eb-bd86-000000000007] stream_transfer_task: Fail to send STREAM_MUTATION_DONE to [BoostrappingNodeIP]:0: std::runtime_error ([Stream #61771151-c554-11eb-bd86-000000000007] GOT STREAM_MUTATION_DONE from [UpNodeIP1]: Can not find stream_manager) WARN
```

Not sure if they are relevant or just different symptoms of the same network problem, but I wanted to share them in case they help you somehow, especially if they point to why the bootstrapping was not automatically retried.

**Comment by FlorianHockmann** [comment]
We also seem to have some performance problems caused by a problem in `iotune`: https://github.com/scylladb/scylla/issues/7146#issuecomment-857670115

Note sure if these could be related or at least explain why the bootstrapping takes so long.

**Comment by FlorianHockmann** [comment]
Unfortunately, the bootstrapping or the following repair resulted in inconsistent data again. Our JanusGraph schema is apparently broken now again.
So, we want to again start with a fresh cluster and the snapshot we created with 4.4.2 on a single node earlier. But this time, we want to directly start with a cluster of 3 nodes and a RF of 3 so we don't have to perform the expensive (and in our case error prone) bootstrapping again. Is the best method to load sstables into a cluster in version 4.4.2 still the `sstableloader` tool or is there a better approach?

**Comment by asias** [comment]
> Unfortunately, the bootstrapping or the following repair resulted in inconsistent data again. Our JanusGraph schema is apparently broken now again.
> So, we want to again start with a fresh cluster and the snapshot we created with 4.4.2 on a single node earlier. But this time, we want to directly start with a cluster of 3 nodes and a RF of 3 so we don't have to perform the expensive (and in our case error prone) bootstrapping again. Is the best method to load sstables into a cluster in version 4.4.2 still the `sstableloader` tool or is there a better approach?


You can try this:

1) start a fresh cluster with 3 empty nodes
2) create the same schema for the user tables
3) copy sstables from snapshot into upload directory on each of the 3 nodes
4) run nodetool refresh myks mytable

**Comment by slivne** [comment]
@FlorianHockmann small comment - make sure to use RF=3 in the schema

**Comment by FlorianHockmann** [comment]
@asias @slivne Just to clarify as we're a bit confused by this note [in the docs](https://docs.scylladb.com/operating-scylla/nodetool-commands/refresh/):

> Scylla node will ignore the partitions in the sstables which are not assigned to this node. For example, if sstable are copied from a different node.

So can we actually load a snapshot that was created from a single node of a 3 node cluster with RF=3 in the 3 nodes of our new cluster with RF=3? The note makes it sound a bit like we would need snapshots from 3 nodes to restore a cluster of 3 nodes.

**Comment by asias** [comment]
> @asias @slivne Just to clarify as we're a bit confused by this note [in the docs](https://docs.scylladb.com/operating-scylla/nodetool-commands/refresh/):
> 
> > Scylla node will ignore the partitions in the sstables which are not assigned to this node. For example, if sstable are copied from a different node.
> 
> So can we actually load a snapshot that was created from a single node of a 3 node cluster with RF=3 in the 3 nodes of our new cluster with RF=3? The note makes it sound a bit like we would need snapshots from 3 nodes to restore a cluster of 3 nodes.

The above procedure is supposed to work.  Since you have RF = 3, the snapshot from one of the 3 nodes will contain ALL possible data.  When you copy the sstables from the snapshot to the empty fresh node3, each of the nodes will think the data belongs to itself and keep the data.

**Comment by FlorianHockmann** [comment]
The procedure worked without any problems so we now have a working cluster again. Thanks a lot for your help on this!
We have also identified a problem in our network that resulted in overloading the Scylla servers with way too many packets and therefore also dropped packages. This probably caused the bootstrapping to fail as @asias already suggested based on the logs.

Regarding why the bootstrapping didn't restart automatically after it failed: We noticed that this only happened for one of our three nodes and this node became completely unresponsive several times in the last few weeks when it had a high load. So this could also explain why Scylla couldn't restart the bootstrapping.
The server is also already 4 years old so we need to replace it any way.

I'm closing this as there was likely no problem in Scylla, but only a network problem and hardware issues on our side.
